docker/Dockerfile

```diff
@@ -42,12 +44,12 @@ RUN set -ex && \
     apt-get update && \
     apt install -y bash tini libc6 libpam-modules krb5-user libnss3 procps && \
     useradd -u ${kyuubi_uid} -g root kyuubi && \
-    mkdir -p ${KYUUBI_HOME} ${KYUUBI_LOG_DIR} ${KYUUBI_PID_DIR} ${KYUUBI_WORK_DIR_ROOT} ${SPARK_HOME} && \
+    mkdir -p ${KYUUBI_HOME} ${KYUUBI_LOG_DIR} ${KYUUBI_PID_DIR} ${KYUUBI_WORK_DIR_ROOT} ${SPARK_BINARY} && \
     chmod ug+rw -R ${KYUUBI_HOME} && \
     chmod a+rwx -R ${KYUUBI_WORK_DIR_ROOT} && \
     rm -rf /var/cache/apt/*
 
-COPY spark-binary ${SPARK_HOME}
+COPY spark-binary ${SPARK_BINARY}
```

Comment:
if user specify the `${spark_home}`, seems we don't need `COPY` the `spark-binary`. So here are tow options:

* user specify the `${spark_home}`; we just do `ENV SPARK_HOME ${spark_home}` without mkdir and `COPY` the `spark-binary`
* user doesn't  specify the `${spark_home}`; we should mkdir and `COPY` the `spark-binary` and `ENV SPARK_HOME SPARK_BINARY`

Reply:
However, Dockerfile seems have not ability to user `if else` to chose copy or not. As I envision it, spark-binary will have no effect on the inherent Spark in the mirror when building with the Spark image as the base image.

Reply:
cc @cfmcgrady do you have good idea ?

Reply:
How about
```Dockerfile
ARG BASE_IMAGE=openjdk:8-jre-slim

ARG BUILD_ENV=copy

FROM ${BASE_IMAGE} as build_copy
ONBUILD RUN echo "copy user specify spark binary...."
ONBUILD COPY user_specify_spark_path /path_to_spark
ONBUILD ENV SPARK_HOME /path_to_spark

FROM ${BASE_IMAGE} as build_no_copy
ONBUILD RUN echo "I don't copy"
ONBUILD ARG BUILDIN_SPARK_HOME="/opt/spark"
ONBUILD ENV SPARK_HOME ${BUILDIN_SPARK_HOME}

FROM build_${BUILD_ENV}
```

- build w/ copy user specify spark binary package

```shell
docker build --build-arg BUILD_ENV=copy -t ${your_image_tag}  -f docker/Dockerfile .
```

- build w/o copy spark binary package
```shell
docker build --build-arg BUILD_ENV=no_copy --build-arg BUILDIN_SPARK_HOME="/path/to/spark_home" -t ${your_image_tag}  -f docker/Dockerfile .
```

Reply:
FYI: https://stackoverflow.com/questions/31528384/conditional-copy-add-in-dockerfile

Reply:
ah, I like the idea using two options of `FROM` + `ONBUILD`, @zwangsheng can you try it ?

Reply:
Nice advice, i will try use `as build`.